magicblock-chainlink/src/remote_account_provider/chain_laser_client.rs

```diff
@@ -0,0 +1,103 @@
+use std::sync::{Arc, Mutex};
+
+use async_trait::async_trait;
+use log::*;
+use solana_pubkey::Pubkey;
+use tokio::sync::{mpsc, oneshot};
+
+use crate::remote_account_provider::{
+    chain_laser_actor::ChainLaserActor,
+    pubsub_common::{ChainPubsubActorMessage, SubscriptionUpdate},
+    ChainPubsubClient, RemoteAccountProviderError, RemoteAccountProviderResult,
+};
+
+#[derive(Clone)]
+pub struct ChainLaserClientImpl {
+    /// Receiver for subscription updates
+    updates: Arc<Mutex<Option<mpsc::Receiver<SubscriptionUpdate>>>>,
+    /// Channel to send messages to the actor
+    messages: mpsc::Sender<ChainPubsubActorMessage>,
+}
+
+impl ChainLaserClientImpl {
+    pub async fn new_from_url(
+        pubsub_url: &str,
+        api_key: &str,
+        commitment: solana_sdk::commitment_config::CommitmentLevel,
+    ) -> RemoteAccountProviderResult<Self> {
+        let (actor, messages, updates) =
+            ChainLaserActor::new_from_url(pubsub_url, api_key, commitment)?;
+        let client = Self {
+            updates: Arc::new(Mutex::new(Some(updates))),
+            messages,
+        };
+        tokio::spawn(actor.run());
+        Ok(client)
+    }
+
+    async fn send_msg(
+        &self,
+        msg: ChainPubsubActorMessage,
+    ) -> RemoteAccountProviderResult<()> {
+        self.messages.send(msg).await.map_err(|err| {
+            RemoteAccountProviderError::ChainLaserActorSendError(
+                err.to_string(),
+                format!("{err:#?}"),
+            )
+        })
+    }
+}
+
+#[async_trait]
+impl ChainPubsubClient for ChainLaserClientImpl {
+    async fn subscribe(
+        &self,
+        pubkey: Pubkey,
+    ) -> RemoteAccountProviderResult<()> {
+        let (tx, rx) = oneshot::channel();
+        self.send_msg(ChainPubsubActorMessage::AccountSubscribe {
+            pubkey,
+            response: tx,
+        })
+        .await?;
+
+        rx.await?
+    }
+
+    async fn unsubscribe(
+        &self,
+        pubkey: Pubkey,
+    ) -> RemoteAccountProviderResult<()> {
+        let (tx, rx) = oneshot::channel();
+        self.send_msg(ChainPubsubActorMessage::AccountUnsubscribe {
+            pubkey,
+            response: tx,
+        })
+        .await?;
+
+        rx.await?
+    }
+
+    async fn shutdown(&self) -> RemoteAccountProviderResult<()> {
+        let (tx, rx) = oneshot::channel();
+        self.send_msg(ChainPubsubActorMessage::Shutdown { response: tx })
+            .await?;
+        rx.await?
+    }
+
+    async fn recycle_connections(&self) {
+        trace!("No need to recycle laserstream connections");
+    }
+
+    fn take_updates(&self) -> mpsc::Receiver<SubscriptionUpdate> {
+        let mut updates_lock = self.updates.lock().unwrap();
+        updates_lock
+            .take()
+            .expect("ChainLaserClientImpl::take_updates called more than once")
+    }
+}
+
+pub fn is_helius_laser_url(url: &str) -> bool {
```

Comment:
I think having a separate config instead of reusing the ws-url is clearer and less confusing. It also leaves room to connect to both in the future for redundancy.